### PR TITLE
combined service discovery

### LINKED ISSUE #111: Combined Service Discovery
add an option to use service discovery information from multiple providers (e.g. Consul and Eureka) for a single service

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #111

### CODE REVIEW

**Comment by michalszynkiewicz** [comment]
cc @cescoffier if you'd like to take a look too

File: pom.xml [hunk 1]
```diff
@@ -242,6 +242,7 @@
         <module>service-discovery/consul</module>
         <module>service-discovery/kubernetes</module>
         <module>service-discovery/eureka</module>
+        <module>service-discovery/combined</module>
```
Comment (cescoffier) [inline pom.xml:245]: I would have named it composite and not combined. What leads you to combine?

File: service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Override
+    public Uni<List<ServiceInstance>> getServiceInstances() {
+        UniJoin.Builder<List<ServiceInstance>> builder = Uni.join().builder();
+        for (ServiceDiscovery element : elements) {
+            builder.add(element.getServiceInstances());
+        }
+
+        return builder.joinAll().andFailFast().onItem().transform(CombiningList::new);
```
Comment (cescoffier) [inline service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java:31]: I'm not sure about the failFast. Maybe that should be a configuration.
We may have the case where you use "combined services" to cover multiple environments. When you run, some service discovery may not be accessible and so fails.

File: service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscoveryProvider.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        String[] services = serviceList.split(",");
+
+        List<String> serviceNames = new ArrayList<>();
+        for (String service : services) {
+            if (service.isBlank()) {
+                throw new IllegalArgumentException("Blank constituent service name for service '" + serviceName + "'");
+            }
+            serviceNames.add(service);
```
Comment (cescoffier) [inline service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscoveryProvider.java:36]: trim and check for blank.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Looks great! 

I've made a few comments, but nothing major. Documentation is also missing.

File: pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply pom.xml:245]: it combines results from other ones :)
composite is good for me too, I'll rename it

Comment (aureamunoz) [reply pom.xml:245]: We should rename all the current `Combined` stuff, no? the classes....

File: service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.groups.UniJoin;
+import io.smallrye.stork.ServiceDiscovery;
+import io.smallrye.stork.ServiceInstance;
+import io.smallrye.stork.Stork;
+import io.smallrye.stork.servicediscovery.combined.util.CombiningList;
+
+public class CombinedServiceDiscovery implements ServiceDiscovery {
```
Comment (aureamunoz) [inline service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java:15]: Rename to `CompositeServiceDiscovery`?

File: service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscoveryProvider.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        return new CombinedServiceDiscovery(serviceNames);
+    }
+
+    @Override
+    public String type() {
+        return "combined";
```
Comment (aureamunoz) [inline service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscoveryProvider.java:43]: rename to `composite`

**Review by aureamunoz — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply pom.xml:245]: yes, that's what I'm doing in an incoming rework :)

File: service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java (hunk 2 above)
Comment (michalszynkiewicz) [reply service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java:31]: There's nothing built in for this, right? The other strategy also returns failure on an error but combines all errors.

Comment (michalszynkiewicz) [reply service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java]: I reworked the other things, I'd rather add support for ignoring a failure later, WDYT @cescoffier ?

Comment (cescoffier) [reply service-discovery/combined/src/main/java/io/smallrye/stork/servicediscovery/combined/CombinedServiceDiscovery.java]: +1

**Review by cescoffier — APPROVED** [review]

File: service-discovery/composite/pom.xml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.smallrye.stork</groupId>
+        <artifactId>smallrye-stork-parent</artifactId>
+        <version>1.0.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>smallrye-stork-service-discovery-combined</artifactId>
```
Comment (aureamunoz) [inline service-discovery/composite/pom.xml:27]: `smallrye-stork-service-discovery-combined` --> `smallrye-stork-service-discovery-composite`

**Review by aureamunoz — CHANGES_REQUESTED** [review]

File: service-discovery/composite/pom.xml (hunk 6 above)
Comment (michalszynkiewicz) [reply service-discovery/composite/pom.xml]: done

**Pull request merged**